### PR TITLE
[Balance] fetch address historical balance

### PR DESCRIPTION
**Pull request by xinyu-li-cb**
### What changed? Why?
Add function to fetch address historical balance

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "license": "ISC",
   "description": "Coinbase Platform SDK",
   "repository": "https://github.com/coinbase/coinbase-sdk-nodejs",
-  "version": "0.0.14",
+  "version": "0.0.12",
```
Comment (erdimaden) [inline package.json:7]: we need to revert this line. it should be 0.0.14>

File: src/client/common.ts [hunk 2]
```diff
@@ -89,17 +89,17 @@ function setFlattenedQueryParams(urlSearchParams: URLSearchParams, parameter: an
     if (typeof parameter === "object") {
         if (Array.isArray(parameter)) {
             (parameter as any[]).forEach(item => setFlattenedQueryParams(urlSearchParams, item, key));
-        }
```
Comment (erdimaden) [inline src/client/common.ts:92]: @xinyu-li-cb why did we change this file?

Comment (xinyu-li-cb) [reply src/client/common.ts:92]: I just copied over all the generated files, likely something wrong with my generator, let me remove it.

Comment (xinyu-li-cb) [reply src/client/common.ts:92]: my ide will remove surfix space...

File: src/tests/historical_balance_test.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { HistoricalBalance } from "../coinbase/historical_balance";
+import { HistoricalBalance as HistoricalBalanceModel } from "../client";
+import { Decimal } from "decimal.js";
+import { Coinbase } from "../coinbase/coinbase";
+
+describe("Balance", () => {
```
Comment (John-peterson-coinbase) [inline src/tests/historical_balance_test.ts:6]:
```suggestion
describe("HistoricalBalance", () => {
```

File: src/tests/historical_balance_test.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  describe(".fromModel", () => {
```
Comment (John-peterson-coinbase) [inline src/tests/historical_balance_test.ts:7]:
```suggestion
  describe("#fromModel", () => {
```

Please prefix static methods with `#methodName` for jest descriptions

File: src/coinbase/address.ts [hunk 5]
```diff
@@ -79,6 +80,41 @@ export class Address {
...
+   * Returns the historical balance of the provided asset.
+   *
+   * @param assetId - The asset ID.
+   * @returns The list of historical balance of the asset.
+   */
+  public async listHistoricalBalance(assetId: string): Promise<HistoricalBalance[]> {
+    const historyList: HistoricalBalance[] = [];
+    const queue: string[] = [""];
+
+    while (queue.length > 0) {
```
Comment (John-peterson-coinbase) [inline src/coinbase/address.ts:93]: does this method list historical balances from genesis?

Should we add paging support so we are not waiting on so many balances before returning the array?

Comment (xinyu-li-cb) [reply src/coinbase/address.ts]: added, thanks for catching!

File: src/coinbase/address.ts [hunk 6]
```diff
@@ -79,6 +80,64 @@ export class Address {
...
 
+  /**
+   * Returns the historical balance of the provided asset.
+   *
+   * @param assetId - The asset ID.
+   * @param limit - A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10.
+   * @param page - A cursor for pagination across multiple pages of results. Don\&#39;t include this parameter on the first call. Use the next_page value returned in a previous response to request subsequent results.
+   * @returns The list of historical balance of the asset and next page token.
+   */
+  public async listHistoricalBalance(
```
Comment (alex-stone) [inline src/coinbase/address.ts:91]: `listHistoricalBalances` since we're returning a list of historical balances

**Pull request closed without merging**